### PR TITLE
[Perf] Optimize DeepSeek-R1 w4afp8 glue kernels

### PR DESCRIPTION
**Pull request by yuhyao**
## Motivation

This PR improves DeepSeek-R1 w4afp8 TP8/EP8 ITL performance. It is motivated by the following profiling trace, obtained by running DeepSeek-R1 w4afp8 TP8 with a concurrency level (i.e., batch size) of 4:

image

The trace reveals several performance gaps:
- The grouped GEMM kernel is a major bottleneck.
- The pre-reorder, post-reorder, and compute_expert_offsets_w4a8 kernels are inefficient.
- There are too many small kernels that should be fused.

## Modifications

Note that this PR does not optimize the w4afp8 grouped GEMM kernel. Although adjusting the tile shape provides some gains, more extensive evaluation is required before making changes to that kernel.

This PR introduces the following optimizations:
- Increased parallelism in the pre-reorder and post-reorder kernels, and fused the multiplication by routed_scaling_factor into the post-reorder kernel.
- Fused the silu_mul and quantization kernels using Triton.
- Replaced the sequential prefix-sum in compute_expert_offsets_w4a8 with a CUB inclusive scan (moved to [PR#12921](https://github.com/sgl-project/sglang/pull/12921)￼).
- Removed unnecessary kernel launches (e.g., torch.zeros, torch.Tensor.to).

After applying these modifications:
image

### TP8
MMLU (sglang bench): 0.869
GSM8K (sglang bench): 0.951
AIME24 (evalscope): 0.8

### EP8
MMLU (sglang bench): 0.870
GSM8K (sglang bench): 0.955
AIME24 (evalscope): 0.8

### Environment
The following results were obtained on 8 × H800 GPUs.

#### TP8
```
python -m sglang.launch_server --model-path path/to/DeepSeek-R1-W4AFP8 --tp-size 8 --context-length 32768 --trust-remote-code --host 0.0.0.0 --port 8000 --mem-fraction-static 0.8 --cuda-graph-max-bs 256 --max-running-requests 256 --random-seed 42 --disable-radix-cache --chunked-prefill-size -1 --cuda-graph-bs 1 2 4 8 16 32 64 128 256
```
#### EP8
```
python -m sglang.launch_server --model-path path/to/DeepSeek-R1-W4AFP8 --tp-size 8 --context-length 32768 --trust-remote-code --host 0.0.0.0 --port 8000 --mem-fraction-static 0.8 --cuda-graph-max-bs 256 --max-running-requests 256 --random-seed 42 --disable-radix-cache --chunked-prefill-size -1 --ep-size 8 --cuda-graph-bs 1 2 4 8 16 32 64 128 256
```

#### TP8, ISL = 10240, OSL = 512, num_prompts=4, rps = inf, max concurrency = 1
Before this PR:
```
============ Serving Benchmark Result ============
Backend:                                 sglang
Traffic request rate:                    inf
Max request concurrency:                 1
Successful requests:                     4
Benchmark duration (s):                  33.31
Total input tokens:                      40960
Total input text tokens:                 40960
Total input vision tokens:               0
Total generated tokens:                  2048
Total generated tokens (retokenized):    2034
Request throughput (req/s):              0.12
Input token throughput (tok/s):          1229.76
Output token throughput (tok/s):         61.49
Total token throughput (tok/s):          1291.25
Concurrency:                             1.00
----------------End-to-End Latency----------------
Mean E2E Latency (ms):                   8321.31
Median E2E Latency (ms):                 8321.94
---------------Time to First Token----------------
Mean TTFT (ms):                          625.81
Median TTFT (ms):                        626.04
P99 TTFT (ms):                           628.17
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          15.06
Median TPOT (ms):                        15.06
P99 TPOT (ms):                           15.07
---------------Inter-Token Latency----------------
Mean ITL (ms):                           15.06
Median ITL (ms):                         15.05
P95 ITL (ms):                            15.19
P99 ITL (ms):                            15.42
Max ITL (ms):                            17.85
==================================================
```

After this PR:
```
============ Serving Benchmark Result ============
Backend:                                 sglang
Traffic request rate:                    inf
Max request concurrency:                 1
Successful requests:                     4
Benchmark duration (s):                  27.28
Total input tokens:                      40960
Total input text tokens:                 40960
Total input vision tokens:               0
Total generated tokens:                  2048
Total generated tokens (retokenized):    2037
Request throughput (req/s):              0.15
Input token throughput (tok/s):          1501.50
Output token throughput (tok/s):         75.08
Total token throughput (tok/s):          1576.58
Concurrency:                             1.00
----------------End-to-End Latency----------------
Mean E2E Latency (ms):                   6815.18
Median E2E Latency (ms):                 6815.67
---------------Time to First Token----------------
Mean TTFT (ms):                          574.46
Median TTFT (ms):                        574.60
P99 TTFT (ms):                           577.04
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          12.21
Median TPOT (ms):                        12.21
P99 TPOT (ms):                           12.21
---------------Inter-Token Latency----------------
Mean ITL (ms):                           12.21
Median ITL (ms):                         12.21
P95 ITL (ms):                            12.37
P99 ITL (ms):                            12.57
Max ITL (ms):                            16.12
==================================================
```

#### TP8, ISL = 10240, OSL = 512, num_prompts=16, rps = inf, max concurrency = 4
Before this PR:
```
============ Serving Benchmark Result ============
Backend:                                 sglang
Traffic request rate:                    inf
Max request concurrency:                 4
Successful requests:                     16
Benchmark duration (s):                  49.67
Total input tokens:                      163840
Total input text tokens:                 163840
Total input vision tokens:               0
Total generated tokens:                  8192
Total generated tokens (retokenized):    8155
Request throughput (req/s):              0.32
Input token throughput (tok/s):          3298.27
Output token throughput (tok/s):         164.91
Total token throughput (tok/s):          3463.19
Concurrency:                             4.00
----------------End-to-End Latency----------------
Mean E2E Latency (ms):                   12409.14
Median E2E Latency (ms):                 12394.34
---------------Time to First Token----------------
Mean TTFT (ms):                          2168.61
Median TTFT (ms):                        2248.19
P99 TTFT (ms):                           2441.57
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          20.04
Median TPOT (ms):                        19.93
P99 TPOT (ms):                           21.45
---------------Inter-Token Latency----------------
Mean ITL (ms):                           20.04
Median ITL (ms):                         19.58
P95 ITL (ms):                            19.80
P99 ITL (ms):                            20.18
Max ITL (ms):                            966.73
==================================================
```

After this PR:
```
============ Serving Benchmark Result ============
Backend:                                 sglang
Traffic request rate:                    inf
Max request concurrency:                 4
Successful requests:                     16
Benchmark duration (s):                  42.68
Total input tokens:                      163840
Total input text tokens:                 163840
Total input vision tokens:               0
Total generated tokens:                  8192
Total generated tokens (retokenized):    8152
Request throughput (req/s):              0.37
Input token throughput (tok/s):          3838.75
Output token throughput (tok/s):         191.94
Total token throughput (tok/s):          4030.68
Concurrency:                             4.00
----------------End-to-End Latency----------------
Mean E2E Latency (ms):                   10661.32
Median E2E Latency (ms):                 10652.05
---------------Time to First Token----------------
Mean TTFT (ms):                          1962.50
Median TTFT (ms):                        2040.72
P99 TTFT (ms):                           2213.48
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          17.02
Median TPOT (ms):                        16.90
P99 TPOT (ms):                           18.32
---------------Inter-Token Latency----------------
Mean ITL (ms):                           17.02
Median ITL (ms):                         16.57
P95 ITL (ms):                            16.82
P99 ITL (ms):                            17.50
Max ITL (ms):                            899.84
==================================================
```

#### TP8, ISL = 1000, OSL = 1000, num_prompts=256, rps = 64, max concurrency = 64
Before this PR:
```
============ Serving Benchmark Result ============
Backend:                                 sglang
Traffic request rate:                    64.0
Max request concurrency:                 64
Successful requests:                     256
Benchmark duration (s):                  178.12
Total input tokens:                      256000
Total input text tokens:                 256000
Total input vision tokens:               0
Total generated tokens:                  256000
Total generated tokens (retokenized):    255280
Request throughput (req/s):              1.44
Input token throughput (tok/s):          1437.23
Output token throughput (tok/s):         1437.23
Total token throughput (tok/s):          2874.46
Concurrency:                             63.81
----------------End-to-End Latency----------------
Mean E2E Latency (ms):                   44400.58
Median E2E Latency (ms):                 44459.86
---------------Time to First Token----------------
Mean TTFT (ms):                          2722.10
Median TTFT (ms):                        2746.66
P99 TTFT (ms):                           3647.66
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          41.72
Median TPOT (ms):                        41.65
P99 TPOT (ms):                           43.61
---------------Inter-Token Latency----------------
Mean ITL (ms):                           41.72
Median ITL (ms):                         41.03
P95 ITL (ms):                            41.79
P99 ITL (ms):                            42.25
Max ITL (ms):                            2957.72
==================================================
```

After this PR:
```
============ Serving Benchmark Result ============
Backend:                                 sglang
Traffic request rate:                    64.0
Max request concurrency:                 64
Successful requests:                     256
Benchmark duration (s):                  159.72
Total input tokens:                      256000
Total input text tokens:                 256000
Total input vision tokens:               0
Total generated tokens:                  256000
Total generated tokens (retokenized):    255308
Request throughput (req/s):              1.60
Input token throughput (tok/s):          1602.85
Output token throughput (tok/s):         1602.85
Total token throughput (tok/s):          3205.69
Concurrency:                             63.80
----------------End-to-End Latency----------------
Mean E2E Latency (ms):                   39801.73
Median E2E Latency (ms):                 39865.50
---------------Time to First Token----------------
Mean TTFT (ms):                          2458.11
Median TTFT (ms):                        2466.51
P99 TTFT (ms):                           3319.97
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          37.38
Median TPOT (ms):                        37.34
P99 TPOT (ms):                           39.06
---------------Inter-Token Latency----------------
Mean ITL (ms):                           37.38
Median ITL (ms):                         36.74
P95 ITL (ms):                            37.56
P99 ITL (ms):                            38.36
Max ITL (ms):                            2566.04
==================================================
```

#### EP8, ISL = 1000, OSL = 1000, num_prompts=256, rps = 64, max concurrency = 64
Before this PR:
```
============ Serving Benchmark Result ============
Backend:                                 sglang
Traffic request rate:                    64.0
Max request concurrency:                 64
Successful requests:                     256
Benchmark duration (s):                  154.72
Total input tokens:                      256000
Total input text tokens:                 256000
Total input vision tokens:               0
Total generated tokens:                  256000
Total generated tokens (retokenized):    255296
Request throughput (req/s):              1.65
Input token throughput (tok/s):          1654.59
Output token throughput (tok/s):         1654.59
Total token throughput (tok/s):          3309.18
Concurrency:                             63.79
----------------End-to-End Latency----------------
Mean E2E Latency (ms):                   38551.11
Median E2E Latency (ms):                 38670.42
---------------Time to First Token----------------
Mean TTFT (ms):                          2649.20
Median TTFT (ms):                        2663.57
P99 TTFT (ms):                           3523.56
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          35.94
Median TPOT (ms):                        35.86
P99 TPOT (ms):                           37.63
---------------Inter-Token Latency----------------
Mean ITL (ms):                           35.94
Median ITL (ms):                         35.19
P95 ITL (ms):                            36.11
P99 ITL (ms):                            37.20
Max ITL (ms):                            2676.41
==================================================
```
After this PR:
```
============ Serving Benchmark Result ============
Backend:                                 sglang
Traffic request rate:                    64.0
Max request concurrency:                 64
Successful requests:                     256
Benchmark duration (s):                  141.63
Total input tokens:                      256000
Total input text tokens:                 256000
Total input vision tokens:               0
Total generated tokens:                  256000
Total generated tokens (retokenized):    255336
Request throughput (req/s):              1.81
Input token throughput (tok/s):          1807.50
Output token throughput (tok/s):         1807.50
Total token throughput (tok/s):          3614.99
Concurrency:                             63.77
----------------End-to-End Latency----------------
Mean E2E Latency (ms):                   35281.71
Median E2E Latency (ms):                 35353.02
---------------Time to First Token----------------
Mean TTFT (ms):                          2360.35
Median TTFT (ms):                        2388.18
P99 TTFT (ms):                           3179.01
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          32.95
Median TPOT (ms):                        32.94
P99 TPOT (ms):                           34.55
---------------Inter-Token Latency----------------
Mean ITL (ms):                           32.96
Median ITL (ms):                         32.30
P95 ITL (ms):                            33.06
P99 ITL (ms):                            33.59
Max ITL (ms):                            2411.77
==================================================
```

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.ai/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.ai/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.ai/developer_guide/contribution_guide.html#write-documentations).
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.ai/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.ai/developer_guide/contribution_guide.html#benchmark-the-speed).

### CODE REVIEW

**Comment by Bruce-x-1997** [comment]
hello，this pr could work now?
and how much performance improvement can it bring?

**Review by FlamingoPg — APPROVED** [review]

**Review by Fridge003 — APPROVED** [review]
